README.md
jsgrowup
========

[![Build Status](https://travis-ci.org/GlobalStrategies/jsgrowup.svg?branch=master)](https://travis-ci.org/GlobalStrategies/jsgrowup.svg?branch=master)

`jsgrowup` by Global Strategies is based on `pygrowup` by Evan Wheeler:
https://github.com/ewheeler/pygrowup

`jsgrowup` calculates z-scores for the following anthropometric indicators:

* weight-for-age

* length/height-for-age

* weight-for-length/height

* head-circumference-for-age

* body-mass-index-for-age

based on the WHO Child Growth Standards:
* http://www.who.int/childgrowth/standards/en/
* http://www.who.int/entity/childgrowth/standards/technical_report/en/index.html

and can optionally use CDC growth standards:
* http://www.cdc.gov/growthcharts

`jsgrowup` avoids floating-point operations to eliminate the unwanted rounding
that muddles the precision of some of the `igrowup` implementations:
* http://docs.sun.com/source/806-3568/ncg_goldberg.html


INSTALLATION
============
`npm install jsgrowup`


EXAMPLE USAGE
=============

`demo.js` includes a sample z-score lookup.


EXCEPTIONS
==========

caller should watch for:

* `AssertionError` raised when caller provides inappropriate parameters

as well as more specific errors (all subclasses of `RuntimeError`):

* `InvalidMeasurement` raised when measurement is invalid for requested indicator

* `InvalidAge` raised when age is invalid for requested indicator

* `DataNotFound` raised when WHO/CDC data is not found for the requested observation (e.g., box-cox, median, coefficient of variance for age)

* `DataError` raised when an error occurs while loading WHO/CDC data into memory


TESTING
=======

* test script using mocha:
`mocha --delay`

* no-framework tests:
`node simpletests.js`

The included tests use example anthropometric data taken from
demonstration data shipped with WHO's `igrowup` software.
`jsgrowup` performs the same calculations and compares the results
to the WHO results.
Please see the sofware licence agreement for WHO's `igrowup`, which
is the source of the test data files:
http://www.who.int/childgrowth/software/license2.pdf

Currently, a small number of cases fail to produce results within 1 standard
deviation of the WHO resuts. These discrepancies may be due to WHO's
use of floating point arithmetic in their `igrowup` software, which leads to
less precise calculations compared to `jsgrowup`. In the absence of any other
trusted test data, please be aware that no claims are made to the
accuracy or reliability of `jsgrowup`'s calculations.


DEVELOPING
==========

The source WHO .txt tables can be easily converted to json with the help of
two amazing python utilities:

* The Pyed Piper https://code.google.com/p/pyp/

* csvkit http://pypi.python.org/pypi/csvkit

heres an example one-liner that changes the source .txt from tsv
to csv (with `pyp`) and then to json (with csvkit's `csvjson`)
`$ cat bmi_girls_2_5_zscores.txt | pyp "p.replace('\t', ',')" | csvjson > bmifa_girls_2_5_zscores.json`


CHANGES
Version 0.8.0 released 2017-09-01
* First release, based on pygrowup


test.js
const vows = require('vows');
const assert = require('assert');
const csv = require('csvtojson');

const grow = require('./jsgrowup');

grow.buildTablesObject(true).then((data) => {
    const batch = {};
    csv()
        .fromFile('./testdata/survey_z_rc.csv')
        .on('json', (rowObj) => {
            function matchesWho(who) {
                const context = {
                    topic: () => {
                        const calc = new grow.Calculator(false, false, data);
                        return calc.zscoreForMeasurement(who.indicator, who.measurement(),
                            who.age, who.gender, who.height);
                    },
                    'within 1.1 SD of WHO': (topic) => {
                        const difference = Math.abs(who.result() - topic);
                        assert.epsilon(1.1, difference, 0);
                    }
                };
                return context;
            }

            [grow.LENGTH_HEIGHT_FOR_AGE, grow.WEIGHT_FOR_LENGTH,
            grow.WEIGHT_FOR_AGE, grow.BODY_MASS_INDEX_FOR_AGE].forEach((ind) => {
                const who = new WhoResult(ind, rowObj);
                // exclude cases with known data problems
                if (who.id === '287') { return; }

                // exclude cases with known discrepancies (still less than 1.6 SD)
                if ((who.id === '136' && ind === grow.LENGTH_HEIGHT_FOR_AGE) ||
                    (who.id === '234' && ind === grow.WEIGHT_FOR_LENGTH) ||
                    (who.id === '234' && ind === grow.BODY_MASS_INDEX_FOR_AGE) || 
                    (who.id === '397' && ind === grow.LENGTH_HEIGHT_FOR_AGE)) { return; } 
                
                // exclude cases with bad data
                if ((ind === grow.LENGTH_HEIGHT_FOR_AGE || ind === grow.WEIGHT_FOR_LENGTH ||
                    ind === grow.WEIGHT_FOR_HEIGHT) && (who.height === null || who.height === '' ||
                        who.height === ' ' || typeof who.height === 'undefined')) {
                    return;
                }
                if (parseInt(who.age, 10) === 0) { return; }
                if ((ind === grow.BODY_MASS_INDEX_FOR_AGE || ind === grow.WEIGHT_FOR_AGE ||
                    ind === grow.WEIGHT_FOR_LENGTH) && who.measurement() === '') { return; }

                    const caseText = `${who.id}: ${who.indicator.toUpperCase()} ` +
                    `(${Math.round(who.measurement() * 10) / 10}) ${who.gender} ` +
                    `age=${who.age} height=${Math.round(who.height * 10) / 10}`;
                batch[caseText] = matchesWho(who);
            });
        })
        .on('end', () => {
            vows.describe('WHO tests').addBatch(batch).run();
        });
});


class WhoResult {
    constructor(indicator, rowObject) {
        this.indicator = indicator;
        Object.keys(rowObject).forEach((key) => {
            this[key.toLowerCase()] = rowObject[key];
        });
        this.age = this.agemons;
        if (parseInt(this.gender, 10) === 1) {
            this.gender = 'M';
        } else if (parseInt(this.gender, 10) === 2) {
            this.gender = 'F';
        } else {
            this.gender = null;
        }
    }

    result() {
        if (this.indicator === grow.LENGTH_HEIGHT_FOR_AGE) {
            return this._zlen;
        }
        if (this.indicator === grow.WEIGHT_FOR_LENGTH) {
            return this._zwfl;
        }
        if (this.indicator === grow.WEIGHT_FOR_AGE) {
            return this._zwei;
        }
        if (this.indicator === grow.BODY_MASS_INDEX_FOR_AGE) {
            return this._zbmi;
        }
    }

    measurement() {
        if (this.indicator === grow.LENGTH_HEIGHT_FOR_AGE) {
            return this.height;
        }
        if ([grow.WEIGHT_FOR_HEIGHT, grow.WEIGHT_FOR_LENGTH].includes(this.indicator)) {
            return this.weight;
        }
        if (this.indicator === grow.WEIGHT_FOR_AGE) {
            return this.weight;
        }
        if (this.indicator === grow.BODY_MASS_INDEX_FOR_AGE) {
            return this._cbmi;
        }
    }

    height() {
        if ([grow.LENGTH_HEIGHT_FOR_AGE, grow.WEIGHT_FOR_HEIGHT,
        grow.WEIGHT_FOR_LENGTH, grow.WEIGHT_FOR_AGE].includes(this.indicator)) {
            return this.height;
        }
    }
}



tests.js
const g = require('./jsgrowup');
const csv = require('csvtojson');

const LONG_OUTPUT = false;
let tables = null;

class WhoResult {
    constructor(indicator, rowObject) {
        this.indicator = indicator;
        Object.keys(rowObject).forEach((key) => {
            this[key.toLowerCase()] = rowObject[key];
        });
        this.age = this.agemons;
        if (parseInt(this.gender, 10) === 1) {
            this.gender = 'M';
        } else if (parseInt(this.gender, 10) === 2) {
            this.gender = 'F';
        } else {
            this.gender = null;
        }
    }

    result() {
        if (this.indicator === g.LENGTH_HEIGHT_FOR_AGE) {
            return this._zlen;
        }
        if (this.indicator === g.WEIGHT_FOR_LENGTH) {
            return this._zwfl;
        }
        if (this.indicator === g.WEIGHT_FOR_AGE) {
            return this._zwei;
        }
        if (this.indicator === g.BODY_MASS_INDEX_FOR_AGE) {
            return this._zbmi;
        }
    }

    measurement() {
        if (this.indicator === g.LENGTH_HEIGHT_FOR_AGE) {
            return this.height;
        }
        if ([g.WEIGHT_FOR_HEIGHT, g.WEIGHT_FOR_LENGTH].includes(this.indicator)) {
            return this.weight;
        }
        if (this.indicator === g.WEIGHT_FOR_AGE) {
            return this.weight;
        }
        if (this.indicator === g.BODY_MASS_INDEX_FOR_AGE) {
            return this.cbmi;
        }
    }

    height() {
        if ([g.LENGTH_HEIGHT_FOR_AGE, g.WEIGHT_FOR_HEIGHT,
        g.WEIGHT_FOR_LENGTH, g.WEIGHT_FOR_AGE].includes(this.indicator)) {
            return this.height;
        }
    }
}

function compareResult(who) {
    let caseText = `${who.id}: ${who.indicator.toUpperCase()} ` +
        `(${Math.round(who.measurement() * 10) / 10}) ${who.gender} ` +
        `age=${who.age} height=${Math.round(who.height * 10) / 10} | `;
    const calc = new g.Calculator(false, false, tables);
    if (who.measurement() && parseFloat(who.age, 10) !== 0) {
        const ourResult = calc.zscoreForMeasurement(who.indicator, who.measurement(),
            who.age, who.gender, who.height);
        caseText += `THEM: ${who.result()} | `;
        if (who.result() && who.result() !== '' && who.result() !== ' ') {
            caseText += `US: ${ourResult} | `;
            const diff = who.result() - ourResult;
            caseText += `DIFF: ${Math.round(Math.abs(diff) * 10) / 10}`;
            const threshold = 1;
            const passed = (Math.abs(diff) < threshold);
            if (LONG_OUTPUT) {
                console.log(`${passed ? 'P' : 'F'} ${caseText}`);
            } else if (passed) {
                process.stdout.write('.');
            } else {
                process.stdout.write('F');
            }
        }
    }
}

g.buildTablesObject(true).then((data) => {
    tables = data;
    csv()
        .fromFile('./testdata/survey_z_rc.csv')
        .on('json', (rowObj) => {
            [g.LENGTH_HEIGHT_FOR_AGE, g.WEIGHT_FOR_LENGTH,
            g.WEIGHT_FOR_AGE, g.BODY_MASS_INDEX_FOR_AGE].forEach((ind) => {
                const who = new WhoResult(ind, rowObj);
                // exclude two cases
                if (who.id === '287' || who.id === '381') { return; }
                if ((ind === g.LENGTH_HEIGHT_FOR_AGE || ind === g.WEIGHT_FOR_LENGTH ||
                    ind === g.WEIGHT_FOR_HEIGHT) && (who.height === null || who.height === '' ||
                        who.height === ' ' || typeof who.height === 'undefined')) {
                    return;
                }
                compareResult(who);
            });
        });
});


test/test.js
const assert = require('chai').assert;
const fs = require('fs');
const Converter = require('csvtojson').Converter;

const grow = require('../index');

grow.buildTablesObject(true).then((data) => {
    const whoCases = [];
    const getWhoCases = () => new Promise((resolve) => {
        const converter = new Converter({});
        const convertToJson = (file) => {
            return new Promise((res, rej) => {
                converter.on('end_parsed', (jsonData) => {
                    if (!jsonData) {
                        rej('CSV to JSON conversion failed!');
                    }
                    res(jsonData);
                });
                fs.createReadStream(file).pipe(converter);
            });
        };

        convertToJson('./testdata/survey_z_rc.csv')
            .then(rows => {
                rows.forEach((rowObj) => {
                    [grow.LENGTH_HEIGHT_FOR_AGE, grow.WEIGHT_FOR_LENGTH,
                    grow.WEIGHT_FOR_AGE, grow.BODY_MASS_INDEX_FOR_AGE].forEach((ind) => {
                        const who = new WhoResult(ind, rowObj);
                        // exclude cases with known data problems
                        if (who.id === '287') { return; }

                        // exclude cases with known discrepancies (still less than 1.6 SD)
                        if ((who.id === '136' && ind === grow.LENGTH_HEIGHT_FOR_AGE) ||
                            (who.id === '234' && ind === grow.WEIGHT_FOR_LENGTH) ||
                            (who.id === '234' && ind === grow.BODY_MASS_INDEX_FOR_AGE) ||
                            (who.id === '397' && ind === grow.LENGTH_HEIGHT_FOR_AGE)) { return; }

                        // exclude cases with bad data
                        if ((ind === grow.LENGTH_HEIGHT_FOR_AGE || ind === grow.WEIGHT_FOR_LENGTH ||
                            ind === grow.WEIGHT_FOR_HEIGHT) && (who.height === null || who.height === '' ||
                                who.height === ' ' || typeof who.height === 'undefined')) {
                            return;
                        }
                        if (parseInt(who.age, 10) === 0) { return; }
                        if ((ind === grow.BODY_MASS_INDEX_FOR_AGE || ind === grow.WEIGHT_FOR_AGE ||
                            ind === grow.WEIGHT_FOR_LENGTH) && who.measurement() === '') { return; }

                        whoCases.push(who);
                    });
                });
                return resolve(whoCases);
            })
            // handle a rejected promise
            .catch(console.warn);
    });
    getWhoCases().then((cases) => {
        describe('Test WHO cases', function () {
            cases.forEach((test) => {
                const caseText = `${test.id}: ${test.indicator.toUpperCase()} ` +
                    `(${Math.round(test.measurement() * 10) / 10}) ${test.gender} ` +
                    `age=${test.age} height=${Math.round(test.height * 10) / 10}`;
                it(`${caseText} within 1.1 SD of WHO`, () => {
                    const calc = new grow.Calculator(false, false, data);
                    const ours = calc.zscoreForMeasurement(test.indicator, test.measurement(),
                        test.age, test.gender, test.height);
                    const difference = Math.abs(test.result() - ours);
                    assert.isAtMost(difference, 1.1);
                });
            });
        });
        run();
    });
});


class WhoResult {
    constructor(indicator, rowObject) {
        this.indicator = indicator;
        Object.keys(rowObject).forEach((key) => {
            this[key.toLowerCase()] = rowObject[key];
        });
        this.age = this.agemons;
        if (parseInt(this.gender, 10) === 1) {
            this.gender = 'M';
        } else if (parseInt(this.gender, 10) === 2) {
            this.gender = 'F';
        } else {
            this.gender = null;
        }
    }

    result() {
        if (this.indicator === grow.LENGTH_HEIGHT_FOR_AGE) {
            return this._zlen;
        }
        if (this.indicator === grow.WEIGHT_FOR_LENGTH) {
            return this._zwfl;
        }
        if (this.indicator === grow.WEIGHT_FOR_AGE) {
            return this._zwei;
        }
        if (this.indicator === grow.BODY_MASS_INDEX_FOR_AGE) {
            return this._zbmi;
        }
    }

    measurement() {
        if (this.indicator === grow.LENGTH_HEIGHT_FOR_AGE) {
            return this.height;
        }
        if ([grow.WEIGHT_FOR_HEIGHT, grow.WEIGHT_FOR_LENGTH].includes(this.indicator)) {
            return this.weight;
        }
        if (this.indicator === grow.WEIGHT_FOR_AGE) {
            return this.weight;
        }
        if (this.indicator === grow.BODY_MASS_INDEX_FOR_AGE) {
            return this._cbmi;
        }
    }

    height() {
        if ([grow.LENGTH_HEIGHT_FOR_AGE, grow.WEIGHT_FOR_HEIGHT,
        grow.WEIGHT_FOR_LENGTH, grow.WEIGHT_FOR_AGE].includes(this.indicator)) {
            return this.height;
        }
    }
}



index.js
'use strict';

// jsgrowup: z-score calculation for anthropometric indicators
// Adapted from pygrowup; BSD license
// Can be used to detect malnutrition, among other applications
// http://www.unhcr.org/45f6abc92.pdf
// Version 0.8.0, following pygrowup

const D = require('decimal.js');
const R = require('ramda');
const fs = require('fs-extra');

const WEIGHT_FOR_LENGTH = 'wfl';
const WEIGHT_FOR_HEIGHT = 'wfh';
const WEIGHT_FOR_AGE = 'wfa';
const LENGTH_HEIGHT_FOR_AGE = 'lhfa';
const HEAD_CIRC_FOR_AGE = 'hcfa';
const BODY_MASS_INDEX_FOR_AGE = 'bmifa';

const AGE_0_2 = '0_2';
const AGE_2_5 = '2_5';
const AGE_0_5 = '0_5';
const AGE_0_13 = '0_13';
const AGE_2_20 = '2_20';

const SEX_BOYS = 'boys';
const SEX_GIRLS = 'girls';

// OBSERVATION CLASS

class Observation {
    constructor(indicator, measurement, ageInMonths, sex, height, american) {
        this.indicator = indicator;
        this.measurement = measurement;
        this.position = null;
        this.ageInMonths = ageInMonths;
        this.sex = sex.toUpperCase();
        this.height = height;
        this.american = american;

        this.tableIndicator = null;
        this.tableAge = null;
        this.tableSex = null;

        if ([WEIGHT_FOR_HEIGHT, WEIGHT_FOR_LENGTH].includes(this.indicator)) {
            if (!height || height === '') {
                throw new Error('no length or height');
            }
        }
    }

    getZScores(calculator) {
        const tableName = this.tableNameForObservation;
        const table = calculator.tables[tableName];
        if ([WEIGHT_FOR_HEIGHT, WEIGHT_FOR_LENGTH].includes(this.indicator)) {
            if (!this.height) { throw new Error('NO HEIGHT'); }
            if (this.height < 45) { throw new Error('too short'); }
            if (this.height > 120) { throw new Error('too tall'); }
            /* find closest height from WHO table (which has data at a resolution
            of half a centimeter). */
            const closestHeight = (Math.round(this.height / 0.5)) * 0.5;
            const scores = table[closestHeight.toString()];
            return scores;
        }
        // all other indicators
        if (this.ageInWeeks <= 13) {
            const closestWeek = Math.floor(this.ageInWeeks);
            const scores = table[closestWeek.toString()];
            return scores;
        }
        // all other ages
        const closestMonth = Math.floor(this.ageInMonths);
        const scores = table[closestMonth.toString()];
        return scores;
    }

    get ageInWeeks() {
        return ((this.ageInMonths * 30.4374) / 7);
    }

    get tableNameForObservation() {
        /* Choose a WHO/CDC table to use, making adjustments
        based on age, length, or height. If, for example, the
        indicator is set to wfl while the child is too long for
        the recumbent tables, this method will make the lookup
        in the wfh table. */

        if (this.sex === 'M') { this.tableSex = SEX_BOYS; }
        if (this.sex === 'F') { this.tableSex = SEX_GIRLS; }

        if ([WEIGHT_FOR_HEIGHT, WEIGHT_FOR_LENGTH].includes(this.indicator)) {
            if (this.indicator === WEIGHT_FOR_LENGTH && this.height > 86) {
                // console.log('too long for recumbent');
                this.tableIndicator = WEIGHT_FOR_HEIGHT;
                this.tableAge = AGE_2_5;
            } else if (this.indicator === WEIGHT_FOR_HEIGHT && this.height < 65) {
                // console.log('too short for standing');
                this.tableIndicator = WEIGHT_FOR_LENGTH;
                this.tableAge = AGE_0_2;
            } else {
                this.tableIndicator = this.indicator;
                if (this.tableIndicator === WEIGHT_FOR_LENGTH) { this.tableAge = AGE_0_2; }
                if (this.tableIndicator === WEIGHT_FOR_HEIGHT) { this.tableAge = AGE_2_5; }
            }
        } else if ([WEIGHT_FOR_AGE, LENGTH_HEIGHT_FOR_AGE, HEAD_CIRC_FOR_AGE].includes(
            this.indicator)) {
            /* weight for age has only one table per sex, as does head circumference for age
            and CDC goes unused before 24mos */
            this.tableIndicator = this.indicator;
            this.tableAge = AGE_0_5;
            if (this.ageInMonths <= 3) {
                if (this.ageInWeeks <= 13) {
                    this.tableAge = AGE_0_13;
                }
            }
            if (this.american && this.ageInMonths >= 24) {
                if (this.indicator === HEAD_CIRC_FOR_AGE) {
                    throw new Error(`TOO OLD ${this.ageInMonths}`);
                }
                this.tableAge = AGE_2_20;
            }
        } else if (this.indicator === BODY_MASS_INDEX_FOR_AGE) {
            this.tableIndicator = this.indicator;
            if (this.ageInMonths <= 3 && this.ageInWeeks <= 13) {
                this.tableAge = AGE_0_13;
            } else if (this.ageInMonths < 24) {
                this.tableAge = AGE_0_2;
            } else if (this.ageInMonths >= 24 && this.ageInMonths <= 60) {
                this.tableAge = AGE_2_5;
            } else if (this.ageInMonths > 60 && this.ageInMonths <= 240) {
                this.tableAge = AGE_2_20;
            } else if (this.ageInMonths > 240) {
                throw new Error(`TOO OLD ${this.ageInMonths}`);
            }
        }

        if (!this.tableIndicator || !this.tableSex || !this.tableAge) {
            throw new Error('DATA ERROR');
        } else {
            const tableName = `${this.tableIndicator}_${this.tableSex}_${this.tableAge}`;
            return tableName;
        }
    }
}

function buildTablesObject(includeCdc = false) {
    /*
    # WHO Growth Standards
    # http://www.who.int/childgrowth/standards/en/
    # WHO tab-separated txt files have been converted to json,
    # and the separate lhfa tables (0-2 and 2-5) have been combined
    */
    const whoTableNames = [
        'wfl_boys_0_2', 'wfl_girls_0_2',
        'wfh_boys_2_5', 'wfh_girls_2_5',
        'lhfa_boys_0_5', 'lhfa_girls_0_5',
        'hcfa_boys_0_5', 'hcfa_girls_0_5',
        'wfa_boys_0_5', 'wfa_girls_0_5',
        'wfa_boys_0_13', 'wfa_girls_0_13',
        'lhfa_boys_0_13', 'lhfa_girls_0_13',
        'hcfa_boys_0_13', 'hcfa_girls_0_13',
        'bmifa_boys_0_13', 'bmifa_girls_0_13',
        'bmifa_boys_0_2', 'bmifa_girls_0_2',
        'bmifa_boys_2_5', 'bmifa_girls_2_5'];

    const getWhoFilePath = tableName => `${__dirname}/tables/${tableName}_zscores.json`;
    const loadWhoFiles = R.pipe(getWhoFilePath, R.pipeP(fs.readFile, JSON.parse));

    const allowedIndexKeys = ['Length', 'Height', 'Month', 'Week'];
    const operativeIndexKey = R.pipe(R.keysIn, R.intersection(allowedIndexKeys), R.head);
    const reIndex = R.indexBy(R.chain(R.prop, operativeIndexKey));
    // find first allowed index key within object keys — this is the operative index key
    // index by the values of the operative index key

    // reindex all
    return Promise.all(R.map(loadWhoFiles, whoTableNames))
        .then(R.map(reIndex))
        .then(R.zipObj(whoTableNames))
        .then((data) => {
            if (includeCdc) {
                /*
                # CDC growth standards
                # http://www.cdc.gov/growthcharts/
                # CDC csv files have been converted to JSON, and the third standard
                # deviation has been fudged for the purpose of this tool.
                */
                const cdcTableNames = [
                    'lhfa_boys_2_20',
                    'lhfa_girls_2_20',
                    'wfa_boys_2_20',
                    'wfa_girls_2_20',
                    'bmifa_boys_2_20',
                    'bmifa_girls_2_20'];

                const getCdcFilePath = tableName => `./tables/${tableName}_zscores.cdc.json`;
                const loadCdcFiles = R.pipe(getCdcFilePath, R.pipeP(fs.readFile, JSON.parse));
                return Promise.all(R.map(loadCdcFiles, cdcTableNames))
                    .then(R.map(reIndex))
                    .then(R.zipObj(cdcTableNames))
                    .then((cdcData) => R.merge(data, cdcData));
            }
            return data;
        });
}

// CALCULATOR CLASS

class Calculator {
    constructor(adjustHeightData = false, adjustWeightScores = false, tables = null) {
        /*
        # Height adjustments are part of the WHO specification
        # (to correct for recumbent vs standing measurements),
        # but none of the existing software seems to implement this.
        # default is false so values are closer to those produced
        # by igrowup software
        */
        this.adjustHeightData = adjustHeightData;

        /*
        # WHO specs include adjustments to z-scores of weight-based
        # indicators that are greater than +/- 3 SDs. These adjustments
        # correct for right skewness and avoid making assumptions about
        # the distribution of data beyond the limits of the observed values.
        # However, when calculating z-scores in a live data collection
        # situation, z-scores greater than +/- 3 SDs are likely to indicate
        # data entry or anthropometric measurement errors and should not
        # be adjusted. Instead, these large z-scores should be used to
        # identify poor data quality and/or entry errors.
        # These z-score adjustments are appropriate only when there
        # is confidence in data quality.
        */
        this.adjustWeightScores = adjustWeightScores;

        this.tables = tables;
        if (!tables) { throw new Error('No data found'); }
    }

    zscoreForMeasurement(indicator, measurement, ageInMonths, sex, height = null,
        american = false) {
        if (!sex || (!['M', 'F', 'm', 'f'].includes(sex))) { throw new Error('Invalid sex value'); }
        if (!ageInMonths || parseInt(ageInMonths, 10) === 0) { throw new Error('Invalid age'); }
        if (!indicator || (![WEIGHT_FOR_LENGTH, WEIGHT_FOR_HEIGHT, LENGTH_HEIGHT_FOR_AGE,
            HEAD_CIRC_FOR_AGE, WEIGHT_FOR_AGE, BODY_MASS_INDEX_FOR_AGE].includes(indicator))) {
            throw new Error('Invalid indicator');
        }
        if (!measurement || parseInt(measurement, 10) === 0) {
            throw new Error('Invalid measurement');
        }

        const obs = new Observation(indicator, measurement, ageInMonths, sex, height, american);
        let m = parseFloat(measurement, 10);

        /*
        # indicator-specific methodology
        # (see section 5.1 of http://www.who.int/entity/childgrowth/standards/\
        #                                  technical_report/en/index.html)
        #
        # TODO accept a recumbent vs standing parameter for deciding
        # whether or not to do these adjustments rather than assuming
        # measurement orientation based on the measurement
        */
        if (indicator === WEIGHT_FOR_LENGTH) {
            if (m > 65.7 && m < 120.7) { m -= 0.7; }
        }
        if (indicator === WEIGHT_FOR_HEIGHT && this.adjustHeightData) {
            m += 0.7;
        }

        const zscores = obs.getZScores(this);
        if (!zscores) { throw new Error('Data error'); }

        const y = new D(m);
        const boxCox = new D(zscores.L);
        const median = new D(zscores.M);
        const coeffVariance = new D(zscores.S);

        /*
        ###
        # calculate z-score
        #
        # (see Chapter 7 of http://www.who.int/entity/childgrowth/standards/\
        #                                  technical_report/en/index.html)
        #
        #           [y/M(t)]^L(t) - 1
        #   Zind =  -----------------
        #               S(t)L(t)
        ###
        */
        const base = y.dividedBy(median);
        // console.log(`${y} ÷ ${median} = ${base}`);
        const power = base.pow(boxCox);
        // console.log(`${y} ** ${boxCox} = ${power}`);
        const numerator = power.minus(1);
        // console.log(`${numerator}`);
        const denominator = coeffVariance.times(boxCox);
        // console.log(`${denominator}`);
        const zscore = numerator.dividedBy(denominator);
        const roundedZscore = zscore.toDecimalPlaces(2).toNumber();
        // console.log(`zscore: ${roundedZscore}`);
        if (!this.adjustWeightScores) {
            return roundedZscore;
        }
        if ([LENGTH_HEIGHT_FOR_AGE, HEAD_CIRC_FOR_AGE,
            BODY_MASS_INDEX_FOR_AGE].includes(indicator)) {
            /*
            # return length/height-for-age (lhfa) without further processing
            # L(t) is always 1 for this indicator, so differences between
            # adjacent SDs (e.g., 2 SD and 3 SD) are constant for a specific
            # age but varied at different ages
            */
            return roundedZscore;
        }
        if (Math.abs(roundedZscore) <= 3) {
            return roundedZscore;
        }
        /*
        # weight-based indicators present right-skewed distributions
        # so use restricted application of LMS method (limiting Box-Cox
        # normal distribution to interval corresponding to z-scores where
        # empirical data are available. z-scores beyond +/- 3 SDs are
        # fixed to the distance between +/- 2 SDs and +/- 3 SD
        # this avoids making assumptions about the distribution of data
        # beyond the limits of observed values
        #
        #            _
        #           |
        #           |       Zind            if |Zind| <= 3
        #           |
        #           |
        #           |       y - SD3pos
        #   Zind* = | 3 + ( ----------- )   if Zind > 3
        #           |         SD23pos
        #           |
        #           |
        #           |
        #           |        y - SD3neg
        #           | -3 + ( ----------- )  if Zind < -3
        #           |          SD23neg
        #           |
        #           |_
        */
        function calcStdDev(sd) {
            /*
            #   e.g.,
            #
            #   SD3neg = M(t)[1 + L(t) * S(t) * (-3)]^ 1/L(t)
            #   SD2pos = M(t)[1 + L(t) * S(t) * (2)]^ 1/L(t)
            #
            ###
            */
            const sdbase = boxCox.times(coeffVariance).times(new D(sd)).plus(1);
            const sdexponent = new D(1).dividedBy(boxCox);
            const sdpower = sdbase.pow(sdexponent);
            return median.times(sdpower);
        }

        if (roundedZscore > 3) {
            console.log('zscore > 3');
            const sd2pos_c = calcStdDev(2);
            const sd3pos_c = calcStdDev(3);

            const sd23dist = sd3pos_c - sd2pos_c;

            //# compute final z-score
            //# zscore = D(3) + ((y - SD3pos_c)/SD23dist)
            const sub = y.minus(sd3pos_c);
            const div = sub.dividedBy(sd23dist);
            const revisedZscore = div.plus(3);
            return revisedZscore.toDecimalPlaces(2).toNumber();
        }

        if (roundedZscore < -3) {
            console.log('zscore < -3');
            const sd2neg_c = calcStdDev(-2);
            const sd3neg_c = calcStdDev(-3);

            const sd23dist = sd2neg_c - sd3neg_c;

            //# compute final z-score
            //# zscore = D(-3) + ((y - SD3neg_c)/SD23dist)
            const sub = y.minus(sd3neg_c);
            const div = sub.dividedBy(sd23dist);
            const revisedZscore = div.minus(3);
            return revisedZscore.toDecimalPlaces(2).toNumber();
        }
    }
}

module.exports = {
    WEIGHT_FOR_LENGTH,
    WEIGHT_FOR_HEIGHT,
    WEIGHT_FOR_AGE,
    LENGTH_HEIGHT_FOR_AGE,
    HEAD_CIRC_FOR_AGE,
    BODY_MASS_INDEX_FOR_AGE,
    AGE_0_2,
    AGE_2_5,
    AGE_0_5,
    AGE_0_13,
    AGE_2_20,
    SEX_BOYS,
    SEX_GIRLS,
    buildTablesObject,
    Calculator,
    Observation
};


simpletests.js
// SIMPLE TESTING (no framework, just console output)

const g = require('./index');
const csv = require('csvtojson');

const LONG_OUTPUT = true;
let tables = null;

class WhoResult {
    constructor(indicator, rowObject) {
        this.indicator = indicator;
        Object.keys(rowObject).forEach((key) => {
            this[key.toLowerCase()] = rowObject[key];
        });
        this.age = this.agemons;
        if (parseInt(this.gender, 10) === 1) {
            this.gender = 'M';
        } else if (parseInt(this.gender, 10) === 2) {
            this.gender = 'F';
        } else {
            this.gender = null;
        }
    }

    result() {
        if (this.indicator === g.LENGTH_HEIGHT_FOR_AGE) {
            return this._zlen;
        }
        if (this.indicator === g.WEIGHT_FOR_LENGTH) {
            return this._zwfl;
        }
        if (this.indicator === g.WEIGHT_FOR_AGE) {
            return this._zwei;
        }
        if (this.indicator === g.BODY_MASS_INDEX_FOR_AGE) {
            return this._zbmi;
        }
    }

    measurement() {
        if (this.indicator === g.LENGTH_HEIGHT_FOR_AGE) {
            return this.height;
        }
        if ([g.WEIGHT_FOR_HEIGHT, g.WEIGHT_FOR_LENGTH].includes(this.indicator)) {
            return this.weight;
        }
        if (this.indicator === g.WEIGHT_FOR_AGE) {
            return this.weight;
        }
        if (this.indicator === g.BODY_MASS_INDEX_FOR_AGE) {
            return this._cbmi;
        }
    }

    height() {
        if ([g.LENGTH_HEIGHT_FOR_AGE, g.WEIGHT_FOR_HEIGHT,
        g.WEIGHT_FOR_LENGTH, g.WEIGHT_FOR_AGE].includes(this.indicator)) {
            return this.height;
        }
    }
}

function compareResult(who) {
    let caseText = `${who.id}: ${who.indicator.toUpperCase()} ` +
        `(${Math.round(who.measurement() * 10) / 10}) ${who.gender} ` +
        `age=${who.age} height=${Math.round(who.height * 10) / 10} | `;
    const calc = new g.Calculator(false, false, tables);
    if (who.measurement() && parseFloat(who.age, 10) !== 0) {
        const ourResult = calc.zscoreForMeasurement(who.indicator, who.measurement(),
            who.age, who.gender, who.height);
        caseText += `THEM: ${who.result()} | `;
        if (who.result() && who.result() !== '' && who.result() !== ' ') {
            caseText += `US: ${ourResult} | `;
            const diff = who.result() - ourResult;
            caseText += `DIFF: ${Math.round(Math.abs(diff) * 10) / 10}`;
            const threshold = 1;
            const passed = (Math.abs(diff) < threshold);
            if (LONG_OUTPUT) {
                console.log(`${passed ? 'P' : 'F'} ${caseText}`);
            } else if (passed) {
                process.stdout.write('.');
            } else {
                process.stdout.write('F');
            }
        }
    }
}

g.buildTablesObject(true).then((data) => {
    tables = data;
    csv()
        .fromFile('./testdata/survey_z_rc.csv')
        .on('json', (rowObj) => {
            [g.LENGTH_HEIGHT_FOR_AGE, g.WEIGHT_FOR_LENGTH,
            g.WEIGHT_FOR_AGE, g.BODY_MASS_INDEX_FOR_AGE].forEach((ind) => {
                const who = new WhoResult(ind, rowObj);
                // exclude two cases
                if (who.id === '287' || who.id === '381') { return; }
                if ((ind === g.LENGTH_HEIGHT_FOR_AGE || ind === g.WEIGHT_FOR_LENGTH ||
                    ind === g.WEIGHT_FOR_HEIGHT) && (who.height === null || who.height === '' ||
                        who.height === ' ' || typeof who.height === 'undefined')) {
                    return;
                }
                compareResult(who);
            });
        });
});
